### ISSUE TITLE
Feature / Issue - File Permission Validation and Verification

### ISSUE THREAD

**Issue by dmblack** [wp-cli/wp-cli#4151]
Hello Team,

Blown away with the wp-cli experience. I'm new to the environment, but have already automated and extended using documentation provided.

I have, however, noticed a pretty critical issue.

File Permissions are definitely not consistent. Additionally; they are broken in some environment configurations. This can be reproduced by downloading and installing wordpress using wp-cli, then removing and reinstalling a theme.

(core download, then core install wordpress)

```wp theme uninstall twentysixteen```
```wp theme install twentysixteen```

The core installation theme permissions on the core install were `drwxrwxr-x`. However, upon re-installation; `drwxr-xr-x`.

This is even more of an issue using custom themes (Such as Bridge) - in which the permissions can be broken entirely, though installation is successful - rendering the site unusable without manual fix.

Arguably; some of the permissions are up to the system / user to manage. However, with the differing permissions in two tool deployment methods (core install, and then theme install) - I think this is something that can be handled better.

I'd like to contribute in the resolution of this issue.

Is there a permissions matrix? I cannot find anything definitive besides user contributed comments. I don't expect a definitive 'xxx' are permissions, as above; these may differ depending on system configuration - but more so a standard on how they are used so far.

**Comment by danielbachhuber** [comment]
Hi @dmblack,

Thanks for starting this conversation.

> Arguably; some of the permissions are up to the system / user to manage. However, with the differing permissions in two tool deployment methods (core install, and then theme install) - I think this is something that can be handled better.

Yes, you are correct. File permission management is the responsibility of the system / user, because appropriate permissions vary between environments and are somewhat subjective.

If WordPress core were to implement permissions enforcement with theme / plugin installation, then WP-CLI would follow suit. WP-CLI simply calls WordPress APIs.

**Comment by dmblack** [comment]
@danielbachhuber
To clarify; does this mean the permissions of files are those as within the archives downloaded from WordPress directly? And that the permissions of the themes inside the core archive then differ from the permissions of the theme archive?

**Comment by danielbachhuber** [comment]
I don't know off the top of my head.

**Comment by dmblack** [comment]
@danielbachhuber 
Thanks for your time so far. Can you point me in the right direction in wp-cli source for;

- how commands are executed against php-cli
- where `wp core download` execution begins

I'm not trying to be lazy, as I have genuinely spent the last hour trying to understand the code base. Unfortunately; I'm not a daily PHP developer - and the abstraction is real! And I certainly don't expect you to do my homework.

Explanation / Reason:
I've compared the files downloaded by wp-cli to those downloaded directly from WordPress The md5sums match, as built into wp-cli. However; the result permissions of manual extraction, and wp-cli `core download` output differ.

So something related only to the `core download` function is resulting in different permissions than desired, likely to the way wp-cli uses php to execute functions - relative to system or user environment configuration.

**Comment by danielbachhuber** [comment]
> how commands are executed against php-cli

I'm not sure I follow. Can you clarify?

> where `wp core download` execution begins

Extraction begins here: https://github.com/wp-cli/core-command/blob/master/src/Core_Command.php#L186

The `Extractor` class itself is here: https://github.com/wp-cli/wp-cli/blob/master/php/WP_CLI/Extractor.php

> So something related only to the `core download` function is resulting in different permissions than desired, likely to the way wp-cli uses php to execute functions - relative to system or user environment configuration.

Interesting. I'm curious to hear what you discover.

**Comment by schlessera** [comment]
@dmblack WP-CLI is run the PHP CLI SAPI, whereas WordPress runs through the PHP Web SAPI (probably php-fpm). On most environments, the web server runs with different permissions than a shell access process, the php-fpm will be run using the web server's permissions.

So, although having different permissions between WordPress and WP-CLI can be problematic, it is probably not something we can avoid.

**Comment by dmblack** [comment]
@schlessera & @danielbachhuber 
Thank you for your responses, and time so far!

Speaking with the core team at WordPress, they were also suggesting this cause. Specifically; they suggested it may be the umask in execution of wp-cli. It's unusual result; the permissions in the core install theme child directories are actually less restrictive than the desired, and that of the actual base archive.

This is why I am confused; as the `wp core download` method is obviously not yet able to call a WordPress API call, right? And when it can call, for the `wp theme install` function - the permissions are as desired.

I should perhaps have been clearer in my original ticket; the tasks I'm working on at the moment are CI and TD development orientated. This is probably the only reason I picked up the change, as the permissions in these directories did not meet my test criteria.

The web service (Apache2) is configured to use php7.1-fcgi (php7.1.5 - run as user).
I also have php7.1-cli installed, I assume this is what wp-cli is using.

I think I've identified where the 'consideration' may be.

Just can't work out now how to rebuild the bin/wp* files.
